mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp

```diff
@@ -894,6 +904,11 @@ struct WarpOpTransferRead : public OpRewritePattern<WarpExecuteOnLane0Op> {
       return failure();
 
     rewriter.replaceAllUsesWith(distributedVal, newRead);
+    if (hasMask) {
+      // Notify the rewriter that the warp op is changing.
+      rewriter.startRootUpdate(warpOp);
```

Comment:
Don't we have to do that before updating the op?

Reply:
We're actually not changing the op in any way. We're only replacing one of its uses with something outside the op, so this is just to notify the rewriter that the op "changed" (because we have some cleanup patterns to drop dead results).

Reply:
We do it at the end to avoid the unfortunate situation where we notify the rewriter of a change despite the pattern failing.

Reply:
Could you put that in the comment? This seems like some spooky-action-at-a-distance, so it would be worth to capture the reasoning.

Reply:
Yeah this should happen earlier. If you want to abort, can call `cancelRootUpdate`.